bin.src/generateCommands.py

```diff
@@ -0,0 +1,581 @@
+
+        print(f"{shellCommand(command)} || return 1")
+
+        ingestCalibs(self.typeName, dataDir, calib, rerun, copyMode)
+
+
+@CalibSource.registerAsSubclass("bias")
+class BiasSource(CalibSource):
+    commandName = "constructPfsBias.py"
+
+@CalibSource.registerAsSubclass("dark")
+class DarkSource(CalibSource):
+    commandName = "constructPfsDark.py"
+
+@CalibSource.registerAsSubclass("flat")
+class FlatSource(CalibSource):
+    commandName = "constructFiberFlat.py"
+
+@CalibSource.registerAsSubclass("fiberTrace")
+class FiberTraceSource(CalibSource):
+    commandName = "constructFiberTrace.py"
+
+    def execute(self, dataDir, calib, rerun, copyMode, processes=1, verbose=False):
+        for v in self.visits:
+            command = [
+                self.commandName,
+                dataDir,
+                f"--calib={os.path.join(dataDir, calib)}",
+                f"--rerun={rerun}",
+                "--id", f"visit={v}",
+                "--batch-type=none", "-j", processes,
+                ]
+
+            if self.configs:
+                command += ["--config"] + self.configs
+
+            print(f"{shellCommand(command)} || return 1")
+
+        fiberTraceDir = os.path.join(dataDir, "rerun", str(rerun), "FIBERTRACE")
+
+        for detector in ["b1", "r1"]:
+            command = textwrap.dedent(fr"""
+                trace0="$(find {shlex.quote(fiberTraceDir)} -maxdepth 1 -name 'pfsFiberTrace-*-{detector}.fits' -print -quit)"
+                if [ -n "$trace0" ]
+                then
+                    mkdir -p {shlex.quote(fiberTraceDir)}/COMBINED
+                    combineFiberTraces.py \
+                        {shlex.quote(fiberTraceDir)}/COMBINED/"$(basename "$trace0")" \
+                        $(find {shlex.quote(fiberTraceDir)} -maxdepth 1 -name 'pfsFiberTrace-*-{detector}.fits' -print) || return 1
+                fi
+            """)
+            print(command[1:].rstrip())
+
+        ingestCalibs("FIBERTRACE/COMBINED", dataDir, calib, rerun, copyMode)
+        #ingestCalibs("DETECTORMAP", dataDir, calib, rerun, copyMode, overwrite=True)
+
+
+@CalibSource.registerAsSubclass("arc")
+class ArcSource(CalibSource):
+    commandName = "reduceArc.py"
+
+    def execute(self, dataDir, calib, rerun, copyMode, processes=1, verbose=False):
+        visits = visitsToString(self.visits)
+
+        command = [
+            self.commandName,
+            dataDir,
+            f"--calib={os.path.join(dataDir, calib)}",
+            f"--rerun={rerun}",
+            "--id", f"visit={visits}",
+            "-j", processes,
+            ]
+
+        if self.configs:
+            command += ["--config"] + self.configs
+
+        print(f"{shellCommand(command)} || return 1")
+
+        #ingestCalibs(self.typeName, dataDir, calib, rerun, copyMode)
+        ingestCalibs("DETECTORMAP", dataDir, calib, rerun, copyMode, overwrite=True)
+
+
+class CalibBlock:
+    calibTypes = [] # will be replaced in processYaml()
```

Comment:
I think this breaks encapsulation. An external function shouldn't be responsible for changing this class' innards.

Reply:
I agree. The current code does not do this.